### PR TITLE
fix: reject top-level fail actions in pipeline validation

### LINKED ISSUE #2021: Reject top-level Fail actions in Extensions
## Problem

The wasm-shim does not support top-level `Fail` operations. When an extension policy includes a `fail` action that doesn't reference a gRPC response variable, it remains top-level. At runtime, the wasm-shim silently skips it:

```
error envoy wasm: Top-level Fail operation is currently unsupported. Action 0: blocked by fail action
```

The request passes through with 200 — the fail action is silently ignored. The policy still shows `Accepted: True` and `Enforced: True`, giving no indication that the action won't work.

## Expected behavior

The extension SDK should reject fail actions that would end up as top-level during pipeline validation (`validateAndAppend`), surfacing `Accepted: False` on the policy status with a clear error message. This protects all extensions using the SDK, not just PipelinePolicy.

### PR DESCRIPTION
**Pull request by crstrn13**
## Summary
- Reject `FailAction`s that don't reference any gRPC response variable in `validateAndAppend()`, since they remain top-level and are silently ignored by the wasm-shim at runtime
- Include previously populated vars (`p.populatedVars`) in `varPatterns` so fail actions referencing gRPC vars from earlier pipeline calls are correctly accepted
- Add comprehensive test cases covering standalone fail, fail with/without gRPC var reference, cross-call references, and no-predicate fail

Closes #2021

## Test plan
- [x] All existing pipeline tests pass
- [x] New `TestPipeline_TopLevelFailAction` subtests cover all edge cases

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Strengthened validation for fail actions: must reference required gRPC response variables and cannot be configured as standalone, top-level, or response-phase actions.
  * Improved error messages clarifying invalid fail-action configurations.
  * Expanded variable-tracking to include variables introduced in the same batch for more accurate validation.

* **Tests**
  * Added and updated tests covering fail-action validation rules and predicate behaviour.

### CODE REVIEW

File: pkg/extension/controller/controller.go [hunk 1]
```diff
@@ -386,17 +386,36 @@ func (p *PipelineImpl) validateAndAppend(phase string, actions []exttypes.Action
...
 		localPopulated[k] = true
 	}
 
 	for _, action := range actions {
+		if _, ok := action.(exttypes.FailAction); ok {
+			refsVar := false
+			for _, expr := range action.CelExpressions() {
+				for _, pattern := range varPatterns {
+					if pattern.MatchString(expr) {
+						refsVar = true
```
Comment (philbrookes) [inline pkg/extension/controller/controller.go:410]: Could you break here?

File: pkg/extension/controller/controller.go [hunk 2]
```diff
@@ -386,17 +386,36 @@ func (p *PipelineImpl) validateAndAppend(phase string, actions []exttypes.Action
...
+					}
+				}
```
Comment (philbrookes) [inline pkg/extension/controller/controller.go:413]: add here:
```suggestion
				}
				if refsVar {
					break
				}
```

**Review by philbrookes — CHANGES_REQUESTED** [review]
looks good, just 2 small suggestions.

File: pkg/extension/controller/controller.go (hunk 1 above)
Comment (crstrn13) [reply pkg/extension/controller/controller.go:410]: correct, didn't notice that. Thanks for the suggestion :+1:

**Review by philbrookes — APPROVED** [review]
